### PR TITLE
US 24 - Optimal Path Journey

### PR DESCRIPTION
**Pull request by GitExcited**
**[For detailed info read the wiki on this PR](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Optional-Feature-(-US%E2%80%9024))**

## Features Added

- Multi-Stop Journey Planning: Implemented a complete end-to-end system that optimizes routes between multiple destinations across campus
- Traveling Salesperson Problem (TSP) Solution: Applied the Nearest Neighbor algorithm to solve the classic TSP challenge of finding efficient routes
- Dual Input Methods: Support for both address-based (via Google Places API) and building/room-based location input
- Task Management: Full task reordering, addition, and removal capabilities with intuitive UI
- Indoor/Outdoor Preferences: Toggle to prioritize covered paths during bad weather conditions
- Tunnel Connection Detection: Visual indication when underground tunnel connections are available between buildings
- Address Preservation: Enhanced outdoor location handling to retain complete address information throughout navigation flow

### Architecture
- Strategy & Decorator Patterns: Implemented combined design patterns to handle diverse route calculations:
- Different strategies for same-building, cross-building, and outdoor routes
- Decorator pattern for composing complex route calculations from reusable steps
- Custom Hooks: Created decoupled state management with `useJourneyPlanner` and `useBuildingRoomSelection` hooks
- Componentization: Built modular UI with 9 distinct components for improved maintainability
### Algorithm

- Nearest Neighbor Implementation: Selected for O(n²) time complexity and practical balance between performance and path quality
- Distance Calculation: Robust calculation system handling complex scenarios:
    - Room-to-room (same floor, same building)
    - Room-to-room (different floor, same building)
    - Room-to-room (different building)
    - Room-to-outdoor location
    - Outdoor-to-outdoor location

### Navigation Enhancements

- Navigation Parameters Standardization: Fixed address formatting for consistent navigation between components
- Room Formatting: Implemented robust room number handling with special cases for building-specific formats
- Tunnel Navigation: Special logic for when to display tunnel options based on user preferences
- Edge Case Handling: Added support for special room types and redundant building prefixes (e.g., H-H109 → H-109)

## Testing

- Algorithm Validation: Created geometric test scenarios with known optimal paths to verify algorithm correctness.
- Indoor-Outdoor Navigation: Verified transitions between indoor and outdoor navigation segments
- Component Testing: Unit tests for UI components and integration tests for the complete planner flow
- Room Format Testing: Verified room formatting logic handles all building-specific formats correctly
- Edge Case Testing: Confirmed the system handles incomplete address data gracefully with fallback mechanisms

## Video

[The link to the youtube video here
](https://www.youtube.com/shorts/fw1xoMsDkCI)

### CODE REVIEW

File: ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    availableFloors,
+    selectedRoom,
+    availableRooms,
+    setSelectedBuilding,
+    setSelectedFloor,
+    setSelectedRoom,
+    resetSelection,
+  } = useBuildingRoomSelection();
+
+  const handleAddressSelect = (location) => {
```
Comment (shanvinluo) [inline ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js:45]: The `handleAddressSelect` and `handleBuildingRoomAdd` callbacks are currently defined inline and contain logic. Since they're only used once but include task and state management, it's appropriate to **keep them extracted** using the **"Extract Function"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA), preserving clarity and enabling easier testing or reuse later.

File: ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+          onMoveDown={moveTaskDown}
+          onRemove={removeTask}
+        />
+
+        <PreferencesSection
+          avoidOutdoor={avoidOutdoor}
+          setAvoidOutdoor={setAvoidOutdoor}
+        />
+
+        <GenerateButton disabled={tasks.length < 2} onPress={generateJourney} />
```
Comment (shanvinluo) [inline ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js:118]: The expression `tasks.length < 2` is directly embedded in the `GenerateButton` prop. To improve clarity and support future reuse or testing, consider computing this condition via a named helper function like `isGenerateDisabled()` using the **"Replace Temp with Query"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA).

File: ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      isPathAllowed: jest.fn(),
+    };
+
+    // Assume createOptimizer returns an object we can modify
+    // Adjust this if createOptimizer has a different structure
+    journeyOptimizer = JourneyOptimizerService.createOptimizer(false); // Assuming this creates the optimizer instance
+    journeyOptimizer.distanceCalculator = mockDistanceCalculator;
+
+    // Mock isPathAllowed to always return true for simplicity in these geometric tests
+    mockDistanceCalculator.isPathAllowed.mockReturnValue(true);
```
Comment (shanvinluo) [inline ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js:20]: The mock behavior for `isPathAllowed` (e.g., always returning true, or conditionally blocking certain nodes) is repeated across multiple tests. Consider creating named mock configurations like `allowAllPaths()` or `disallowSpecificNodes()` using the **"Extract Function"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA). This will improve clarity and reusability of the test setup.

File: ConcordiaMaps/hooks/useBuildingRoomSelection.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  const [availableRooms, setAvailableRooms] = useState([]);
+
+  // Load buildings on mount
+  useEffect(() => {
+    setBuildings(FloorRegistry.getBuildings());
+  }, []);
+
+  // Update available floors when building changes
+  useEffect(() => {
+    if (selectedBuilding) {
```
Comment (shanvinluo) [inline ConcordiaMaps/hooks/useBuildingRoomSelection.js:19]: The repeated logic for deriving the `buildingType` from `selectedBuilding` in both `useEffect` hooks could be extracted into a shared helper like `getBuildingType()` using the **"Extract Function"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA). This will reduce duplication and clarify intent.

File: ConcordiaMaps/hooks/useBuildingRoomSelection.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      console.log("Getting available floors from building");
+      const buildingType =
+        FloorRegistry.getBuildingTypeFromId(selectedBuilding);
+      if (buildingType) {
+        const building = FloorRegistry.getBuilding(buildingType);
+        if (building && building.floors) {
+          setAvailableFloors(Object.keys(building.floors));
+          console.log("Getting available floors from building");
```
Comment (shanvinluo) [inline ConcordiaMaps/hooks/useBuildingRoomSelection.js:27]: The `console.log` statements in production code should be removed or replaced with a proper logging mechanism using the **"Remove Dead Code"** or **"Inline Temp"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA), especially if no longer needed.

File: ConcordiaMaps/services/JourneyOptimizer/RouteDecorators.js [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  };
+};
+
+export const BaseCalculation = () => () => {
+  // Base calculation returns 0 as the starting point
+  return 0;
+};
+
+// Helper function to normalize building IDs
+const normalizeBuildingId = (buildingId) => {
```
Comment (shanvinluo) [inline ConcordiaMaps/services/JourneyOptimizer/RouteDecorators.js:224]: The logic for normalizing building IDs is embedded in the same file but could be reused elsewhere. Consider moving `normalizeBuildingId()` into a shared utility module like `BuildingUtils` using the **"Move Function"** refactoring pattern from the [Catalog](https://github.com/SOEN-390-pmzone/SOEN-_390-Q-QD/wiki/Refactoring-Template-for-QA) to improve reusability and separation of concerns.

**Review by shanvinluo — CHANGES_REQUESTED** [review]
Absolutely incredible work David! Except for a few refactoring opportunities, I don't see anything I would change or add. Please just improve the test coverage for SonarQube.
I tested all functionalities manually and they work:

- Multi-stop journey planning runs smoothly with optimized routes
- Address and building/room inputs are both functional
- Task management (add, reorder, delete) works without issues
- Indoor/outdoor toggle correctly affects routing
- Tunnel detection and address preservation are reliable
- Route calculations are accurate across all scenarios tested

Also verified edge cases, room formatting logic, and fallback behaviors — all good! 

So, please just address the refactoring issues and improve code coverage

File: ConcordiaMaps/components/JourneyPlanner/LocationsSection.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import { View, Text } from "react-native";
+import PropTypes from "prop-types";
+import LocationsList from "./LocationsList";
+import styles from "../../styles/JourneyPlanner/JourneyPlannerScreenStyles";
+
+const LocationsSection = ({ tasks, onMoveUp, onMoveDown, onRemove }) => {
+  return (
+    <View style={styles.tasksContainer}>
+      <Text style={styles.sectionTitle}>Your Locations ({tasks.length})</Text>
+      {tasks.length === 0 ? (
```
Comment (GolimarGit) [inline ConcordiaMaps/components/JourneyPlanner/LocationsSection.js:11]: Just to make sure, you're checking that there are **at least 2** locations?

File: ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js (hunk 1 above)
Comment (GitExcited) [reply ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js:45]: Yes normally this would be a good refactoring; however 

1. The functions are already defined outside the JSX return statement
2. Each function has a single responsibility
3. The functions already improve redability and make test easier

So I would say that for this case the refactoring isn't necessary. Let me know if you disagree or if you think its still worth doing.

File: ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js (hunk 3 above)
Comment (GitExcited) [reply ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js:20]: Yes great comment. This would improve readability and maintainability. I will perform the extract function refactoring. You can see it in commit d4f3dabd968f1aa0f18d1016bcc6937cfdd46e82

File: ConcordiaMaps/hooks/useBuildingRoomSelection.js (hunk 4 above)
Comment (GitExcited) [reply ConcordiaMaps/hooks/useBuildingRoomSelection.js:24]: Yes you are right about this. The function would allow both components to use this helper function following the **Extract Function** refactoring pattern. You can see the changed code in commit 77411bce1af6a2ade036ef6127896aaf9e9ecdde

File: ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js (hunk 2 above)
Comment (GitExcited) [reply ConcordiaMaps/components/JourneyPlanner/JourneyPlannerScreen.js]: Yes great comment. This would improve readability and maintainability. I will perform the extract function refactoring. You can see it in commit d4f3dabd968f1aa0f18d1016bcc6937cfdd46e82

**Comment by GitExcited** [comment]
After doing some refactoring, it seems like some tests are failing. Ill fix it then commit again

File: ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js (hunk 3 above)
Comment (GitExcited) [reply ConcordiaMaps/__tests__/services/JourneyOptimizer/JourneyOptimizerService.test.js]: That's a great idea. i've reworked this test to not only extract that function but others that were being reused like `useEuclideanDistance`, `allowPaths`, `disallowSpecificNodes`. You can see the changes on this commit: bc4ccc565199712004783da39f4ae2905367df2a

File: ConcordiaMaps/hooks/useBuildingRoomSelection.js (hunk 5 above)
Comment (GitExcited) [reply ConcordiaMaps/hooks/useBuildingRoomSelection.js:31]: Thanks! Yes I removed the console.log print.

File: ConcordiaMaps/services/JourneyOptimizer/RouteDecorators.js (hunk 6 above)
Comment (GitExcited) [reply ConcordiaMaps/services/JourneyOptimizer/RouteDecorators.js:224]: I appreciate the suggestion, however this particular refactoring might be uncesessary since the function is:

- Relatively small (15 lines including comments)
- Currently used only within this file
- Well-documented with a clear purpose
- Properly placed at the end of the file with appropriate comments

Should this function be needed elsewhere in the future, we can certainly move it at that time. For now, its scope is appropriately contained within the routing decorators domain.

File: ConcordiaMaps/components/JourneyPlanner/LocationsSection.js (hunk 7 above)
Comment (GitExcited) [reply ConcordiaMaps/components/JourneyPlanner/LocationsSection.js:11]: Yes, the check for having at least 2 locations is implemented in the `JourneyPlannerScreen` component, not in this `LocationsSection` component.

**Comment by JeremyVieira** [comment]
Tested building to building, classroom to classroom and the mix combination of both. Properly optimizes the path between the chosen locations. Also tested the tunnel and is working as it indicates to use the tunnel when possible.

**Review by JeremyVieira — APPROVED** [review]

**Comment by GitExcited** [comment]
test failing because of prettier

**Review by JeremyVieira — APPROVED** [review]
I've reviewed the comprehensive multi-stop journey planning implementation and I'm impressed with the breadth and depth of the changes. This feature set represents a significant enhancement to our navigation capabilities.

Strong Points
Architecture: The use of Strategy and Decorator patterns provides a clean separation of concerns for the different navigation scenarios. This makes the code more maintainable and extensible.
Algorithm Implementation: The Nearest Neighbor approach for TSP offers a good balance of performance and solution quality for our campus navigation needs.
User Experience: The dual input methods (address-based and building/room-based) give users flexibility in how they plan their journeys.
Edge Case Handling: I appreciate the attention to detail in handling special room formats and building-specific conventions.
Technical Implementation
The code demonstrates thoughtful architecture decisions:

The componentization into 9 distinct modules improves maintainability
The custom hooks for state management keep components clean and focused
The standardization of navigation parameters helps maintain consistency across different navigation types
Testing Coverage
The testing approach is thorough, covering:

Algorithm validation with geometric test scenarios
Indoor-outdoor navigation transitions
Component-level unit tests
Room formatting edge cases
Graceful handling of incomplete data
Recommendation
Based on the comprehensive implementation, thorough testing, and careful architectural decisions, I recommend merging this PR. The multi-stop journey planning feature will significantly enhance the navigation experience for users across campus.

**Comment by mannrp** [comment]
Since most of the conflicts/refactoring was done while merging US15.3, and Code Smells have been resolved for the most part, and sufficient coverage is reached. It looks good to merge.

**Review by mannrp — APPROVED** [review]

**Review by IdrisDrouiche — APPROVED** [review]

File: ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+  /**
+   * Calculates the distance between two locations using the appropriate strategy
+   * @param {Object} locationA - First location
+   * @param {Object} locationB - Second location
+   * @returns {number} Distance between the locations
+   */
+  calculateDistance(locationA, locationB) {
+    console.log(
+      "DistanceCalculatorService: Dtermining the distance between A and B",
```
Comment (IdrisDrouiche) [inline ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js:38]: Type here. Pretty small and unsignificant change, but promotes proper code hygiene.

**Review by GolimarGit — APPROVED** [review]

File: ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js (hunk 8 above)
Comment (JeremyVieira) [reply ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js]: fixed the typo

File: ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+  /**
+   * Determines the appropriate strategy based on location types
+   * @private
+   * @param {Object} locationA - First location
+   * @param {Object} locationB - Second location
+   * @returns {Object} Appropriate strategy object
+   */
+  _determineStrategy(locationA, locationB) {
+    // Validate input parameters
+    if (!this._areLocationsValid(locationA, locationB)) {
```
Comment (IdrisDrouiche) [inline ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js:60]: Could there be a use case where we want to return a "null strategy" instead of defaulting to outdoor? For example, to signal completely invalid input?

Comment (GitExcited) [reply ConcordiaMaps/services/JourneyOptimizer/DistanceCalculatorService.js:60]: Yes that's a good idea. However, we already have a check in place insie `JourneyOptimizer` that checks for invalid inputs. Therefore, this function doesn't need to have the extra strategy. 

However, in the future it could be interesting to add more strategies like a Tunnel one
